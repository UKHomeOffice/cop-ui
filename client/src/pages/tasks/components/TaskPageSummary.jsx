import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';
import { Link } from 'react-navi';
import moment from 'moment';
import ChangePriority from './ChangePriority';

const TaskPageSummary = ({
  businessKey,
  category,
  assignee,
  taskInfo,
  taskUpdateSubmitted,
  setTaskUpdateSubmitted,
}) => {
  const { t } = useTranslation();
  const [isEditingPriority, setIsEditingPriority] = useState(false);
  const handlePriorityEdit = () => {
    setIsEditingPriority(!isEditingPriority);
  };

  return (
    <>
      <div className="govuk-grid-row">
        <div className="govuk-grid-column-full" id="taskName">
          <span className="govuk-caption-l">
            <Link
              className="govuk-link"
              target="_blank"
              rel="noopener noreferrer"
              href={`/cases/${businessKey}`}
            >
              {businessKey}
            </Link>
          </span>
          <h2 className="govuk-heading-l">{taskInfo.name}</h2>
        </div>
      </div>
      <div className="govuk-grid-row">
        <div className="govuk-grid-column-one-quarter" id="category">
          <span className="govuk-caption-m govuk-!-font-size-19">{t('pages.task.category')}</span>
          <h4 className="govuk-heading-m govuk-!-font-size-19">{category}</h4>
        </div>
        <div className="govuk-grid-column-one-quarter" id="taskDueDate">
          <span className="govuk-caption-m govuk-!-font-size-19">{t('pages.task.due')}</span>
          <h4 className="govuk-heading-m govuk-!-font-size-19">
            {moment().to(moment(taskInfo.due))}
          </h4>
        </div>
        <div className="govuk-grid-column-one-quarter" id="taskPriority">
          <span className="govuk-caption-m govuk-!-font-size-19">
            {t(`pages.task.priority`)}
            &nbsp; (
            <button
              className="govuk-accordion__open-all"
              aria-hidden="true"
              type="button"
              onClick={handlePriorityEdit}
              onKeyDown={handlePriorityEdit}
            >
              {isEditingPriority ? 'cancel' : 'change'}
            </button>
            )
            <ChangePriority
              isEditingPriority={isEditingPriority}
              taskInfo={taskInfo}
              taskUpdateSubmitted={taskUpdateSubmitted}
              setTaskUpdateSubmitted={setTaskUpdateSubmitted}
            />
          </span>
        </div>
        <div className="govuk-grid-column-one-quarter" id="taskAssignee">
          <span className="govuk-caption-m govuk-!-font-size-19">{t('pages.task.assignee')}</span>
          <h4 className="govuk-heading-m govuk-!-font-size-19">{assignee}</h4>
        </div>
      </div>
      <div className="govuk-grid-row">
        <div className="govuk-grid-column-full" id="description">
          <p className="govuk-body">{taskInfo.description}</p>
        </div>
      </div>
    </>
  );
};

TaskPageSummary.defaultProps = {
  assignee: '',
  taskInfo: {
    description: '',
  },
};

TaskPageSummary.propTypes = {
  businessKey: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired,
  assignee: PropTypes.string,
  taskInfo: PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
    due: PropTypes.string.isRequired,
  }),
  taskUpdateSubmitted: PropTypes.bool.isRequired,
  setTaskUpdateSubmitted: PropTypes.func.isRequired,
};

export default TaskPageSummary;
