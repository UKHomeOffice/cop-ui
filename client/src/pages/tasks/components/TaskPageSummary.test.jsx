import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import TaskPageSummary from './TaskPageSummary';

describe('TaskPageSummary', () => {
  const mockAxios = new MockAdapter(axios);
  const mockSetTaskUpdateSubmitted = jest.fn();
  const mockProps = {
    businessKey: 'businessKey',
    category: 'test',
    assignee: 'assignee',
    taskInfo: {
      id: 'id',
      priority: 100,
      name: 'name',
      due: '2021-03-19T12:06:03.964+0000',
      description: 'description',
    },
    taskUpdateSubmitted: false,
    setTaskUpdateSubmitted: mockSetTaskUpdateSubmitted,
  };

  beforeEach(() => {
    mockAxios.reset();
  });

  it('can render without error', () => {
    render(<TaskPageSummary {...mockProps} />);
  });

  it('can toggle priority editing', () => {
    render(<TaskPageSummary {...mockProps} />);

    fireEvent.click(screen.getAllByText('change')[0]);

    expect(screen.queryAllByText('cancel')[0]).toBeInTheDocument();
    expect(screen.queryByText('Change priority')).toBeInTheDocument();
    expect(screen.queryByText('Low')).toBeInTheDocument();
    expect(screen.queryByText('Medium')).toBeInTheDocument();
    expect(screen.queryByText('High')).toBeInTheDocument();

    fireEvent.click(screen.getAllByText('cancel')[0]);

    expect(screen.getAllByText('change')[0]).toBeInTheDocument();
    expect(screen.queryByText('Change priority')).not.toBeInTheDocument();
    expect(screen.queryByText('Low')).not.toBeInTheDocument();
    expect(screen.queryByText('Medium')).toBeInTheDocument();
    expect(screen.queryByText('High')).not.toBeInTheDocument();
  });
});
