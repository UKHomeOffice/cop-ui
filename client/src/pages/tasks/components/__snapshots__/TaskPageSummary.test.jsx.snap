// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskPageSummary matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-full"
      id="taskName"
    >
      <span
        class="govuk-caption-l"
      >
        businessKey
      </span>
      <h1
        class="govuk-heading-l"
      >
        name
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-quarter"
      id="category"
    >
      <span
        class="govuk-caption-m govuk-!-font-size-19"
      >
        pages.task.category
      </span>
      <h4
        class="govuk-heading-m govuk-!-font-size-19"
      >
        test
      </h4>
    </div>
    <div
      class="govuk-grid-column-one-quarter"
      id="taskDueDate"
    >
      <span
        class="govuk-caption-m govuk-!-font-size-19"
      >
        pages.task.due  (
        <button
          aria-hidden="true"
          class="govuk-accordion__open-all"
          type="button"
        >
          change
        </button>
        )
      </span>
      <h4
        class="govuk-heading-m govuk-!-font-size-19"
      >
        <span
          aria-label="due 3 months ago"
          class="govuk-!-font-size-19 govuk-!-font-weight-bold overdue"
        >
          Overdue 3 months ago
        </span>
      </h4>
    </div>
    <div
      class="govuk-grid-column-one-quarter"
      id="taskPriority"
    >
      <span
        class="govuk-caption-m govuk-!-font-size-19"
      >
        pages.task.priority  (
        <button
          aria-hidden="true"
          class="govuk-accordion__open-all"
          type="button"
        >
          change
        </button>
        )
      </span>
      <h4
        class="govuk-heading-m govuk-!-font-size-19"
      >
        Medium
      </h4>
    </div>
    <div
      class="govuk-grid-column-one-quarter"
      id="taskAssignee"
    >
      <span
        class="govuk-caption-m govuk-!-font-size-19"
      >
        pages.task.assignee
      </span>
      <h4
        class="govuk-heading-m govuk-!-font-size-19"
      >
        pages.task.unassigned
      </h4>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-full"
      id="description"
    >
      <p
        class="govuk-body"
      >
        description
      </p>
    </div>
  </div>
</DocumentFragment>
`;
